db/db_batch.go

```diff
@@ -120,125 +139,96 @@ func (b *baseKVStoreBatch) Entry(index int) (*writeInfo, error) {
-}
-
-// Commit persists pending writes to DB and clears the write queue
-func (b *boltDBBatch) Commit() error {
-	var err error
-	numRetries := b.bdb.config.NumRetries
-	for c := uint8(0); c < numRetries; c++ {
-		err = b.bdb.db.Update(func(tx *bolt.Tx) error {
-			for _, write := range b.writeQueue {
-				if write.writeType == Put {
-					bucket, err := tx.CreateBucketIfNotExists([]byte(write.namespace))
-					if err != nil {
-						return errors.Wrapf(err, write.errorFormat, write.errorArgs)
-					}
-					if err := bucket.Put(write.key, write.value); err != nil {
-						return errors.Wrapf(err, write.errorFormat, write.errorArgs)
-					}
-				} else if write.writeType == PutIfNotExists {
-					bucket, err := tx.CreateBucketIfNotExists([]byte(write.namespace))
-					if err != nil {
-						return errors.Wrapf(err, write.errorFormat, write.errorArgs)
-					}
-					if bucket.Get(write.key) == nil {
-						if err := bucket.Put(write.key, write.value); err != nil {
-							return errors.Wrapf(err, write.errorFormat, write.errorArgs)
-						}
-					} else {
-						return ErrAlreadyExist
-					}
-				} else if write.writeType == Delete {
-					bucket := tx.Bucket([]byte(write.namespace))
-					if bucket == nil {
-						return nil
-					}
-					if err := bucket.Delete(write.key); err != nil {
-						return errors.Wrapf(err, write.errorFormat, write.errorArgs)
-					}
-				}
-			}
-
-			return nil
-		})
-
-		if err == nil || err == ErrAlreadyExist {
-			break
-		}
+// PutIfNotExists inserts a <key, value> record only if it does not exist yet, otherwise return ErrAlreadyExist
+func (cb *cachedBatch) PutIfNotExists(namespace string, key, value []byte, errorFormat string, errorArgs ...interface{}) error {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+	if _, ok := cb.cache[cb.hash(namespace, key)]; ok {
+		return ErrAlreadyExist
 	}
+	cb.cache[cb.hash(namespace, key)] = value
+	return cb.batch(PutIfNotExists, namespace, key, value, errorFormat, errorArgs)
+}
 
-	// clear queues
-	b.Clear()
+// Delete deletes a record
+func (cb *cachedBatch) Delete(namespace string, key []byte, errorFormat string, errorArgs ...interface{}) error {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+	delete(cb.cache, cb.hash(namespace, key))
+	return cb.batch(Delete, namespace, key, nil, errorFormat, errorArgs)
+}
+
+// Clear clear the cached batch buffer
+func (cb *cachedBatch) Clear() error {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+	return cb.clear()
+}
+
+// Get retrieves a record
+func (cb *cachedBatch) Get(namespace string, key []byte) ([]byte, error) {
+	cb.mutex.RLock()
+	defer cb.mutex.RUnlock()
+	if v, ok := cb.cache[cb.hash(namespace, key)]; ok {
+		return v, nil
+	}
+	return nil, ErrNotExist
+}
 
-	return err
+//======================================
+// private functions
+//======================================
+func (cb *cachedBatch) hash(namespace string, key []byte) hash.PKHash {
+	stream := []byte(namespace)
+	stream = append(stream, key...)
```

Comment:
Add dot between namespace and key, otherwise it's not safe concat. e.g.

namespace = "foo", key = []byte("bar")
namespace = "fo", key = []byte("obar")

Reply:
it should be using the same namespace
but good catch 👍 will modify